ibm-etag-header crashes linter on incorrect specification

Hello,
I'm using the latest version of openapi-validator (IBM OpenAPI Validator (validator: 1.2.0; ruleset: 1.2.0)).

On invalid specification
```yml
openapi: 3.0.3

paths:
  /test:
    get:
      operationId: ruleBug
      parameters:
        - $ref: '#/components/schemas/NotAParameter'
      responses:
        '204':
          description: No content
components:
  schemas:
    NotAParameter:
      type: string
```

rule `ibm-etag-header` crashes whole linter:

```
[ERROR] There was a problem with spectral.
[ERROR] Error running Nimma
[ERROR] Additional error details:
[ERROR] AggregateError: Error running Nimma
    at Scope.destroy (/opt/linters/node_modules/nimma/dist/legacy/cjs/runtime/scope.js:137:13)
    at Nimma.eval (eval at query (/opt/linters/node_modules/nimma/dist/legacy/cjs/core/index.js:66:71), <anonymous>:182:11)
    at Nimma.query (/opt/linters/node_modules/nimma/dist/legacy/cjs/core/index.js:68:71)
    at execute (/opt/linters/node_modules/@stoplight/spectral-core/dist/runner/runner.js:82:11)
    at Runner.run (/opt/linters/node_modules/@stoplight/spectral-core/dist/runner/runner.js:51:13)
    at Spectral.runWithResolved (/opt/linters/node_modules/@stoplight/spectral-core/dist/spectral.js:55:22)
    at async Spectral.run (/opt/linters/node_modules/@stoplight/spectral-core/dist/spectral.js:63:17)
    at async runSpectral (/opt/linters/node_modules/ibm-openapi-validator/src/spectral/spectral-validator.js:39:27)
    at async runValidator (/opt/linters/node_modules/ibm-openapi-validator/src/cli-validator/run-validator.js:201:17) {
  [errors]: [
    RuntimeError: $.paths[*] threw: ErrorWithCause("Function \"\" threw an exception: Cannot read properties of undefined (reading 'toLowerCase')")
        at _callbacks.<computed> (/opt/linters/node_modules/nimma/dist/legacy/cjs/runtime/proxy-callbacks.js:34:21)
        at Scope.emit (/opt/linters/node_modules/nimma/dist/legacy/cjs/runtime/scope.js:109:19)
        ... 4 lines matching cause stack trace ...
        at _traverseBody (/opt/linters/node_modules/nimma/dist/legacy/cjs/runtime/traverse.js:17:5)
        at _traverse (/opt/linters/node_modules/nimma/dist/legacy/cjs/runtime/traverse.js:41:7)
        at Scope.traverse (/opt/linters/node_modules/nimma/dist/legacy/cjs/runtime/traverse.js:47:3)
        at Scope.traverse (/opt/linters/node_modules/nimma/dist/legacy/cjs/runtime/scope.js:94:25) {
      cause: [ErrorWithCause]
    }
  ]
}
```

Side issue – I wonder if it would be possible to output better error message containing rule name and perhaps rule input? I have bunch of rules specific to my projects, and it can be time-consuming to figure out which rule has error.

--- a/packages/ruleset/src/functions/array-of-arrays.js
+++ b/packages/ruleset/src/functions/array-of-arrays.js
@@ -20,20 +20,20 @@ module.exports = function (schema, _opts, context) {
 };
 
 function arrayOfArrays(schema, path) {
-  const errors = [];
-
   if (isArraySchema(schema) && schema.items) {
     logger.debug(
       `${ruleId}: checking array schema at location: ${path.join('.')}`
     );
     if (isArraySchema(schema.items)) {
       logger.debug('Found an array of arrays!');
-      errors.push({
-        message: 'Array schemas should avoid having items of type array',
-        path,
-      });
+      return [
+        {
+          message: 'Array schemas should avoid having items of type array',
+          path,
+        },
+      ];
     }
   }
 
-  return errors;
+  return [];
 }

--- a/packages/ruleset/src/functions/etag-header-exists.js
+++ b/packages/ruleset/src/functions/etag-header-exists.js
@@ -8,7 +8,7 @@ const { LoggerFactory } = require('../utils');
 let ruleId;
 let logger;
 
-module.exports = function (pathItem, options, context) {
+module.exports = function etagHeaderExists(pathItem, options, context) {
   if (!logger) {
     ruleId = context.rule.name;
     logger = LoggerFactory.getInstance().getLogger(ruleId);
@@ -138,9 +138,9 @@ function isETagNeeded(pathItem) {
 function headerParamsPresent(paramList, headerNames) {
   if (Array.isArray(paramList)) {
     for (const p of paramList) {
-      if (p.in.toLowerCase() === 'header') {
+      if (p.in && p.in.toLowerCase() === 'header') {
         for (const headerName of headerNames) {
-          if (headerName.toLowerCase() === p.name.toLowerCase()) {
+          if (p.name && headerName.toLowerCase() === p.name.toLowerCase()) {
             return true;
           }
         }

--- a/packages/validator/src/cli-validator/run-validator.js
+++ b/packages/validator/src/cli-validator/run-validator.js
@@ -200,9 +200,7 @@ async function runValidator(cliArgs, parseOptions = {}) {
     try {
       results = await runSpectral({ validFile, originalFile }, context);
     } catch (err) {
-      logError('There was a problem with spectral.', getError(err));
-      logger.error('Additional error details:');
-      logger.error(err);
+      handleSpectralError(err);
       exitCode = 1;
       continue;
     }
@@ -251,6 +249,47 @@ function logError(description, message = '') {
   logger.error(`${description}`);
   if (message) {
     logger.error(`${message}`);
+  }
+}
+
+function handleSpectralError(error) {
+  logError('There was a problem with spectral.', getError(error));
+  logger.error('Additional error details:');
+
+  // The nimma errors are especially difficult to understand, so we do some parsing
+  // based on the expected structure of the errors to extract the error cause and
+  // the specific file/line number the problem occurs on.
+  if (
+    error.message &&
+    error.message === 'Error running Nimma' &&
+    error instanceof AggregateError
+  ) {
+    const errorDedupMap = {};
+    error.errors.forEach(err => {
+      if (err.cause && err.cause.cause) {
+        const { cause } = err.cause;
+        try {
+          // Look for the filepath/line number at the top of the stack.
+          const topOfStack = cause.stack.split(' at ')[1];
+          const reason = cause.message;
+          if (errorDedupMap[reason + topOfStack]) {
+            // Don't print duplicates, continue the loop.
+            return;
+          }
+          logger.error(`Cause: ${reason}`);
+          logger.error(`At: ${topOfStack}`);
+          errorDedupMap[reason + topOfStack] = true;
+        } catch (e) {
+          logger.debug('Could not parse Spectral error information');
+          logger.debug(e.message);
+          logger.error(cause);
+        }
+      } else {
+        logger.error(error);
+      }
+    });
+  } else {
+    logger.error(error);
   }
 }
 

